### DISCUSSION TITLE
Crazy edge case question

### DISCUSSION THREAD

**Discussion by Waxhaw** [michaelrsweet/pappl#109] [General]
What if I wanted to build an AirPrint product and I already had my own custom print engine?

For example...

I just want to advertise a generic printer to do photo printing for an AirPrint interface for iOS and Android.

Would this be difficult?

So currently I'm doing this via a generic ppd file and a custom filter to move the image to a 'hot folder' for printing.

Based on the C API code examples, this looks like a single solution that doesn't make it easy to open that part up.

Curious what other think.  I would love to get away from legacy ppd files.

TIA!

**Comment by michaelrsweet** [comment]
@Waxhaw So PAPPL provides the IPP printer semantics/wrapper, plus some basic job processing/raster printing functionality. If you have your own file processing code, you can probably integrate it fairly easily - the only thing you'll need to do is support the raster printing interface for clients that send PWG/Apple Raster.

**Reply by Waxhaw** [reply]
What if I wanted to just use your framework for transport?

So something like 'AirDrop' or 'AirPrint' and I use a 'hot folder' type approach?

I've already built everything else and it works well.

What I don't have is an easy way to 'advertise a service' so it works with any SmartPhone which you handle very well.

**Reply by michaelrsweet** [reply]
You'd need to write the basics of a PAPPL driver to communicate the capabilities of your service. You'd also need to be prepared to receive raster data (which is what many clients will send you instead of PDF) and convert that to something your code can support.

**Reply by michaelrsweet** [reply]
I should also note that you'd never be able to get AIrPrint certification for such a solution. There needs to be a way for users to cancel jobs and monitor status, not just blindly print-and-pray...

**Reply by Waxhaw** [reply]
I'm wanting to build something like this:  https://www.amazon.com/DNP-WPS-Pro-Wireless-Server/dp/B0719BVJ69

Minus the touch screen and a very specific printer and vertical market.

A dedicated photo printer with it's own directly attached print server that makes printing photos off a phone as simple and easy as possible.

I have everything working now with CUPS but I'm trying to get away from it.  I'm sure you, more than anyone else on the planet knows the many reasons why.  :-)

Cheers!

**Reply by michaelrsweet** [reply]
Well, without more information it is really hard to advise you further, but I really (*really*) think you will want to do something more than a hot folder. PAPPL can absolutely be used to provide AirPrint compatibility and if it handles the dispatching of jobs to the printer than you get all of the things you'd need to actually get certified. If you just use it to dump images into a hot folder then it won't give you the level of control I think you'll want.

**Reply by Waxhaw** [reply]
It's photo printing only.  Dedicated printer.  4x6 or a 'fixed' image size.  There is nothing for the user to control.  The printer will be right in front of them.

Simple is as simple does.  ;-)
